# Livro para consulta:
- https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html
- https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html
    

# 1. Importando bibliotecas <a name="import"></a>

<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline

# 2. Carregando o dataframe SINASC <a name="read"></a>
<div style="text-align: right"
     
[Voltar ao índice](#Contents)

In [3]:
sinasc_raw = pd.read_csv('SINASC_RO_2019.csv')
sinasc_raw.head()

ORIGEM   CODESTAB  CODMUNNASC  LOCNASC  IDADEMAE  ESTCIVMAE  \
0       1  2679477.0      110001        1        19        5.0   
1       1  2679477.0      110001        1        29        2.0   
2       1  2679477.0      110001        1        37        9.0   
3       1  2516500.0      110001        1        30        5.0   
4       1  2516500.0      110001        1        30        2.0   

            ESCMAE  CODOCUPMAE  QTDFILVIVO  QTDFILMORT  ...  KOTELCHUCK  \
0      8 a 11 anos         NaN         0.0         0.0  ...           5   
1      8 a 11 anos    999992.0         1.0         0.0  ...           5   
2      8 a 11 anos    513205.0         2.0         0.0  ...           5   
3  12 anos ou mais    231205.0         0.0         0.0  ...           4   
4      8 a 11 anos    999992.0         1.0         0.0  ...           5   

  CONTADOR munResStatus munResTipo               munResNome  munResUf  \
0        1        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
1        2        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
2        3        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   
3        4        ATIVO      MUNIC  Alto Alegre dos Parecis  Rondônia   
4        5        ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia   

   munResLat munResLon  munResAlt  munResArea  
0  -11.93554 -61.99982      338.0    7067.025  
1  -11.93554 -61.99982      338.0    7067.025  
2  -11.93554 -61.99982      338.0    7067.025  
3  -12.13178 -61.85308      397.0    3958.273  
4  -11.93554 -61.99982      338.0    7067.025  

[5 rows x 69 columns]

In [5]:
sinasc_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27028 entries, 0 to 27027
Data columns (total 69 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ORIGEM        27028 non-null  int64  
 1   CODESTAB      26913 non-null  float64
 2   CODMUNNASC    27028 non-null  int64  
 3   LOCNASC       27028 non-null  int64  
 4   IDADEMAE      27028 non-null  int64  
 5   ESTCIVMAE     26711 non-null  float64
 6   ESCMAE        26716 non-null  object 
 7   CODOCUPMAE    24121 non-null  float64
 8   QTDFILVIVO    25455 non-null  float64
 9   QTDFILMORT    24930 non-null  float64
 10  CODMUNRES     27028 non-null  int64  
 11  GESTACAO      25796 non-null  object 
 12  GRAVIDEZ      26949 non-null  object 
 13  PARTO         26979 non-null  object 
 14  CONSULTAS     27028 non-null  int64  
 15  DTNASC        27028 non-null  object 
 16  HORANASC      27007 non-null  float64
 17  SEXO          27024 non-null  object 
 18  APGAR1        26932 non-nu

# Tarefa 1

### 1. Idade media das mães e dos pais por município (coluna munResNome)


In [8]:
media_1 = sinasc_raw[['IDADEMAE', 'IDADEPAI', 'munResNome']]
media_1.dtypes

IDADEMAE        int64
IDADEPAI      float64
munResNome     object
dtype: object

In [11]:
media_1.pivot_table(index='munResNome')

IDADEMAE   IDADEPAI
munResNome                                     
Alta Floresta D'Oeste      25.991826  29.548701
Alto Alegre dos Parecis    24.844156  29.179104
Alto Paraíso               24.960177  28.833333
Alvorada D'Oeste           25.770732  30.757282
Ariquemes                  25.607866  32.466667
Buritis                    25.559177  30.812500
Cabixi                     26.037500  34.333333
Cacaulândia                25.546667  36.333333
Cacoal                     26.890102  30.931854
Campo Novo de Rondônia     24.825301  30.500000
Candeias do Jamari         25.243323  29.894737
Castanheiras               27.281250  30.392857
Cerejeiras                 27.240506  31.652778
Chupinguaia                25.333333  29.629630
Colorado do Oeste          27.602804  34.354839
Corumbiara                 24.716535  32.470588
Costa Marques              24.412017  30.613636
Cujubim                    24.482927  31.400000
Espigão D'Oeste            26.093333  31.198113
Governador Jorge Teixeira  24.401786  29.222222
Guajará-Mirim              25.462228  30.470588
Itapuã do Oeste            26.064220  42.500000
Jaru                       25.982974  32.197183
Ji-Paraná                  26.361137  31.132239
Machadinho D'Oeste         25.480000  31.326923
Ministro Andreazza         27.904762  33.849315
Mirante da Serra           26.304878  32.154762
Monte Negro                25.648241  34.678571
Município ignorado - RO    24.000000  22.000000
Nova Brasilândia D'Oeste   26.222222  30.520833
Nova Mamoré                25.326389  31.692308
Nova União                 26.894231  31.371429
Novo Horizonte do Oeste    26.938776  31.588235
Ouro Preto do Oeste        26.375594  31.233244
Parecis                    26.454545  32.473684
Pimenta Bueno              25.199675  29.745174
Pimenteiras do Oeste       25.075000  35.600000
Porto Velho                26.312670  32.352679
Presidente Médici          25.796053  30.150754
Primavera de Rondônia      27.023256  31.888889
Rio Crespo                 26.660000  28.000000
Rolim de Moura             26.600863  30.360479
Santa Luzia D'Oeste        26.121212  30.703297
Seringueiras               24.910714  30.500000
São Felipe D'Oeste         26.481481  31.395833
São Francisco do Guaporé   25.660066  31.638889
São Miguel do Guaporé      25.926346  32.666667
Teixeirópolis              26.031250  31.500000
Theobroma                  25.536082  30.083333
Urupá                      26.688235  31.008929
Vale do Anari              26.096774  34.000000
Vale do Paraíso            25.679012  29.976744
Vilhena                    26.433962  34.152941

### 2. Peso médio dos bebes por sexo que nasceram no dia do seu aniversário por faixas de escolaridade mae
Ex: Você, aluna(o), nasceu no dia 10/01, então você precisa filtrar o conjunto de dados nessa data e calcular o peso médio dos bebês de cada sexo por faixa de escolaridade da mãe.

In [44]:
media_2 = sinasc_raw[['PESO', 'ESCMAE', 'DTNASC','SEXO']]
media_2.head(5)

PESO           ESCMAE      DTNASC       SEXO
0  3685      8 a 11 anos  2019-02-19  Masculino
1  3055      8 a 11 anos  2019-02-21   Feminino
2  3460      8 a 11 anos  2019-02-25   Feminino
3  2982  12 anos ou mais  2019-03-20   Feminino
4  3310      8 a 11 anos  2019-03-23   Feminino

In [57]:
md1 = media_2[media_2['DTNASC'].isin(['2019-11-02'])]
md1.head(20)

PESO           ESCMAE      DTNASC       SEXO
14705  3850      8 a 11 anos  2019-11-02  Masculino
14780  4270       1 a 3 anos  2019-11-02  Masculino
14781  3620  12 anos ou mais  2019-11-02  Masculino
14782  2940       4 a 7 anos  2019-11-02  Masculino
14783  3705      8 a 11 anos  2019-11-02  Masculino
14790  3650              NaN  2019-11-02  Masculino
15726  2280      8 a 11 anos  2019-11-02   Feminino
15727  2620      8 a 11 anos  2019-11-02  Masculino
16593  2740      8 a 11 anos  2019-11-02   Feminino
16594  3760      8 a 11 anos  2019-11-02  Masculino
16595  3060      8 a 11 anos  2019-11-02   Feminino
16628  3285       4 a 7 anos  2019-11-02  Masculino
17071  3135      8 a 11 anos  2019-11-02   Feminino
17072  3295      8 a 11 anos  2019-11-02   Feminino
17073  3280      8 a 11 anos  2019-11-02   Feminino
17163  2495       4 a 7 anos  2019-11-02   Feminino
18024  3700      8 a 11 anos  2019-11-02  Masculino
18025  4140       4 a 7 anos  2019-11-02   Feminino
18037  3930       4 a 7 anos  2019-11-02   Feminino
19164  4010      8 a 11 anos  2019-11-02   Feminino

In [85]:
md1.groupby(['ESCMAE', 'SEXO'])['PESO'].mean().unstack()

SEXO                Feminino    Masculino
ESCMAE                                   
1 a 3 anos               NaN  3752.500000
12 anos ou mais  3247.500000  2715.000000
4 a 7 anos       3492.000000  3071.875000
8 a 11 anos      3036.470588  3556.363636

### 3. Qual o municipio que nasceu menos bebe em 2019?
    - qual a idade media, maxima, minima das maes nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?

In [171]:
# preparando dataframe filtrado
menos_bebe = sinasc_raw[['IDADEMAE','IDADEPAI','munResNome','QTDFILVIVO']]
menos_bebe

IDADEMAE  IDADEPAI               munResNome  QTDFILVIVO
0            19      26.0    Alta Floresta D'Oeste         0.0
1            29      24.0    Alta Floresta D'Oeste         1.0
2            37      32.0    Alta Floresta D'Oeste         2.0
3            30      24.0  Alto Alegre dos Parecis         0.0
4            30      27.0    Alta Floresta D'Oeste         1.0
...         ...       ...                      ...         ...
27023        32      35.0                  Vilhena         1.0
27024        19       NaN              Chupinguaia         0.0
27025        24      30.0                  Vilhena         0.0
27026        21      24.0                  Vilhena         1.0
27027        18       NaN  Novo Horizonte do Oeste         1.0

[27028 rows x 4 columns]

In [104]:
# municipio que nasceu menos bebe em 2019

menos_bebe.groupby(['munResNome'])['QTDFILVIVO'].count().sort_values(ascending=True)[:5]


munResNome
Município ignorado - RO     1
Castanheiras               32
Pimenteiras do Oeste       40
Primavera de Rondônia      43
Parecis                    43
Name: QTDFILVIVO, dtype: int64

O município com menos bebes nascidos vivos no ano de 2019 é 'Castanheiras' com 32 bebes. O 'Município ignorado - RO' apresenta uma possível falta de informações, pois é mostrado um resultado de apenas 1 bebe nascido vivo no ano de 2019.

In [114]:
# idade media, maxima, minima das MAES/PAIS nesse municipio
menos_bebe[menos_bebe['munResNome'] == 'Castanheiras'].groupby('munResNome').aggregate(['min', np.median, max]).drop(columns=['QTDFILVIVO'])


IDADEMAE            IDADEPAI             
                  min median max      min median   max
munResNome                                            
Castanheiras       17   27.0  39     17.0   31.5  43.0

### 4. Qual o municipio que nasceu mais bebe no mês de março?
    - qual a quantidade de filhos vivos media, maxima, minima nesse municipio?
    - qual a idade media, maxima, minima dos pais nesse municipio?



In [165]:
# transformando o DTNASC em datetime

sinasc_raw['DTNASC'] = pd.to_datetime(sinasc_raw['DTNASC'])

# criando coluna mês, para facilitar analise

sinasc_raw['MÊS'] = pd.DatetimeIndex(sinasc_raw['DTNASC']).month

In [178]:
# preparando dataframe filtrado

mais_bebe = sinasc_raw[['IDADEPAI','munResNome','QTDFILVIVO', 'MÊS']]

mais_bebe.head(12)


IDADEPAI               munResNome  QTDFILVIVO  MÊS
0       26.0    Alta Floresta D'Oeste         0.0    2
1       24.0    Alta Floresta D'Oeste         1.0    2
2       32.0    Alta Floresta D'Oeste         2.0    2
3       24.0  Alto Alegre dos Parecis         0.0    3
4       27.0    Alta Floresta D'Oeste         1.0    3
5       40.0    Alta Floresta D'Oeste         2.0    3
6       42.0  Novo Horizonte do Oeste         1.0    4
7       27.0    Alta Floresta D'Oeste         0.0    4
8       25.0    Alta Floresta D'Oeste         0.0    2
9       41.0    Alta Floresta D'Oeste         1.0    3
10      28.0    Alta Floresta D'Oeste         1.0    3
11      29.0    Alta Floresta D'Oeste         1.0    3

In [182]:
# municipio que nasceu mais bebe no mês de março

mais_bebe[mais_bebe['MÊS'] == 3].groupby(['munResNome'])['QTDFILVIVO'].count().sort_values(ascending=False)[:5]


munResNome
Porto Velho    627
Ji-Paraná      187
Vilhena        147
Ariquemes      139
Cacoal         132
Name: QTDFILVIVO, dtype: int64

O município com mais bebes nascidos vivos no ano de 2019 é 'Porto Velho' com 627 bebes.

In [185]:
# quantidade de filhos vivos media, maxima, minima nesse municipio

mais_bebe[mais_bebe['munResNome'] == 'Porto Velho'].groupby('munResNome').aggregate([np.median,max,'min']).drop(columns=['MÊS','IDADEPAI'])


QTDFILVIVO           
                median   max  min
munResNome                       
Porto Velho        1.0  12.0  0.0

In [188]:
# idade media, maxima, minima dos pais nesse municipio

mais_bebe[mais_bebe['munResNome'] == 'Porto Velho'].groupby('munResNome').aggregate([np.median,max,'min']).drop(columns=['MÊS','QTDFILVIVO'])


IDADEPAI            
              median   max   min
munResNome                      
Porto Velho     32.0  65.0  16.0

### Analise as respostas encontradas, tire algum insight delas, conte pra gente algo encontrado nos dados. Algo que você julgue relevante e novo pra você.

Percebi que os municípios com mais bebes nascidos são aqueles onde o pai em geral é maior de 30 anos.

In [6]:
sinasc_raw.dtypes

ORIGEM          int64
CODESTAB      float64
CODMUNNASC      int64
LOCNASC         int64
IDADEMAE        int64
               ...   
munResUf       object
munResLat     float64
munResLon     float64
munResAlt     float64
munResArea    float64
Length: 69, dtype: object

In [7]:
sinasc_raw.select_dtypes(exclude=['int64','float64'])

ESCMAE         GESTACAO GRAVIDEZ    PARTO      DTNASC  \
0          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-02-19   
1          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-02-21   
2          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-02-25   
3      12 anos ou mais  37 a 41 semanas    Única  Cesáreo  2019-03-20   
4          8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-03-23   
...                ...              ...      ...      ...         ...   
27023  12 anos ou mais  32 a 36 semanas    Única  Cesáreo  2019-12-13   
27024      8 a 11 anos  37 a 41 semanas    Única  Vaginal  2019-10-04   
27025      8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-08-02   
27026      8 a 11 anos  32 a 36 semanas    Única  Vaginal  2019-12-23   
27027      8 a 11 anos  37 a 41 semanas    Única  Cesáreo  2019-12-27   

            SEXO RACACOR  DTCADASTRO CODANOMAL VERSAOSIST   DTRECEBIM  \
0      Masculino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
1       Feminino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
2       Feminino  Branca  2019-03-08       NaN     3.2.01  2019-03-08   
3       Feminino   Parda  2019-04-03       NaN     3.2.01  2019-04-09   
4       Feminino   Parda  2019-04-03       NaN     3.2.01  2019-04-09   
...          ...     ...         ...       ...        ...         ...   
27023  Masculino  Branca  2020-09-28       NaN     3.2.01  2020-10-13   
27024  Masculino  Branca  2019-10-16       NaN     3.2.01  2019-10-21   
27025  Masculino   Parda  2019-08-07       NaN     3.2.00  2019-08-07   
27026  Masculino   Parda  2020-01-07       NaN     3.2.00  2020-01-07   
27027  Masculino   Parda  2020-10-15       NaN     3.2.01  2020-10-16   

      munResStatus munResTipo               munResNome  munResUf  
0            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
1            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
2            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
3            ATIVO      MUNIC  Alto Alegre dos Parecis  Rondônia  
4            ATIVO      MUNIC    Alta Floresta D'Oeste  Rondônia  
...            ...        ...                      ...       ...  
27023        ATIVO      MUNIC                  Vilhena  Rondônia  
27024        ATIVO      MUNIC              Chupinguaia  Rondônia  
27025        ATIVO      MUNIC                  Vilhena  Rondônia  
27026        ATIVO      MUNIC                  Vilhena  Rondônia  
27027        ATIVO      MUNIC  Novo Horizonte do Oeste  Rondônia  

[27028 rows x 15 columns]

In [4]:
sinasc_raw.columns

Index(['ORIGEM', 'CODESTAB', 'CODMUNNASC', 'LOCNASC', 'IDADEMAE', 'ESTCIVMAE',
       'ESCMAE', 'CODOCUPMAE', 'QTDFILVIVO', 'QTDFILMORT', 'CODMUNRES',
       'GESTACAO', 'GRAVIDEZ', 'PARTO', 'CONSULTAS', 'DTNASC', 'HORANASC',
       'SEXO', 'APGAR1', 'APGAR5', 'RACACOR', 'PESO', 'IDANOMAL', 'DTCADASTRO',
       'CODANOMAL', 'NUMEROLOTE', 'VERSAOSIST', 'DTRECEBIM', 'DIFDATA',
       'DTRECORIGA', 'NATURALMAE', 'CODMUNNATU', 'CODUFNATU', 'ESCMAE2010',
       'SERIESCMAE', 'DTNASCMAE', 'RACACORMAE', 'QTDGESTANT', 'QTDPARTNOR',
       'QTDPARTCES', 'IDADEPAI', 'DTULTMENST', 'SEMAGESTAC', 'TPMETESTIM',
       'CONSPRENAT', 'MESPRENAT', 'TPAPRESENT', 'STTRABPART', 'STCESPARTO',
       'TPNASCASSI', 'TPFUNCRESP', 'TPDOCRESP', 'DTDECLARAC', 'ESCMAEAGR1',
       'STDNEPIDEM', 'STDNNOVA', 'CODPAISRES', 'TPROBSON', 'PARIDADE',
       'KOTELCHUCK', 'CONTADOR', 'munResStatus', 'munResTipo', 'munResNome',
       'munResUf', 'munResLat', 'munResLon', 'munResAlt', 'munResArea'],
      dtype='object')